Python in mantid (#28806)
mantidproject/mantid: docs/source/tutorials/python_in_mantid/further_alg_ws/06_extract_manipulate_data.rst

.. _06_extract_manipulate_data:

=====================================
Extract and Manipulate Data: Examples
=====================================

Read vs Extract
===============

* *Read* produces a view into the chosen part of the original data.
* *Extract* creates a copy of this part of the data.

Read
----

Using a loop, read and print the first value in all spectra

.. code-block:: python

	for index in range(0, raw_workspace.getNumberHistograms()):
	    #Note the round brackets followed by the square brackets
	    print(raw_workspace.readY(index)[0])

Workspace data can be read as numpy arrays, spectrum by spectrum:

.. code-block:: python

	ws = Load(Filename="HRP39182.RAW")
	for i in range(ws.getNumberHistograms()):
	    y = ws.readY(i)
	    x = ws.readX(i)
	    e = ws.readE(i)

Warning
-------

> **Be careful**: the outputs of *read* (y,x,e) are only **views into the data held by the workspace,** `ws`. If `ws` is deleted, the contents of x,y,e will be nonsense (the random contents of the memory locations formerly used for `ws`).

Reviewer: ```suggestion
**Be careful**: the outputs of *read* (y,x,e) are only **views into the data held by the workspace,** `ws`. If `ws` is deleted, the contents of x,y,e will be invalid (the random contents of the memory locations formerly used for `ws`).
```

> If you need x,y,e to persist longer than ws, use the *extract*, which creates a copy of the data in `ws` into y,x,e.

Reviewer: ```suggestion
If you need x,y,e data to persist longer than the workspace use the *extract* methods, which create a copy of the data in `ws`.
```

Extract
-------

The data from all spectra can be obtained as a mutable multi-dimensional array in one-call using the extract methods.

.. code-block:: python

	ws = Load(Filename="HRP39182.RAW")
	x = ws.extractX()

	print(x.shape)

> 	print(y.shape)

Reviewer: Are these print statements correct, as we don't have a `y` or `e` variable, or maybe I'm missing something.

Author: I'd copied that over from here and for some reason left out the extractY and extractE lines!

	print(e.shape)

Since the *extract* methods return multi-dimensional numpy arrays. So to use  *extract* in a similar way to *read*, you need to `slice these arrays with indexing <https://numpy.org/doc/1.18/reference/arrays.indexing.html>`_. 

E.g. instead of `ws.readX(5)` you should use:

.. code-block:: python

    ws.extractX()[5, :]

    xmat = ws.extractX(); x = xmat[5, :]

Nested Looping
==============

> This allows access to the individual bins in each spectrum. e.g. To sum the y-values in each spectrum:

Reviewer: ```suggestion
This allows access to the individual bins in each spectrum. E.g. to sum the y-values in each spectrum:
```
